File: R/densify.R
`%||%` <- function(a, b) if (is.null(a)) b else a

#' Adaptively densify lines for a projection
#'
#' Insert vertices along great circle arcs exactly where a target projection
#' bends them, so that straight chords drawn between output vertices stay
#' within `tolerance` of the true projected curve. This is the adaptive
#' resampling scheme of d3-geo (Bostock), generalized to paths and to
#' segment meshes.
#'
#' The error metric is evaluated in the plane of the `target` projection:
#' the great circle midpoint of each candidate segment is forward-projected
#' and compared to the segment's planar chord (perpendicular distance, plus
#' a parametric check that catches strong shear). Segments are split
#' recursively until they pass or `max_depth` is reached. No inverse
#' transforms are used and each point is projected exactly once.
#'
#' Coordinates are returned in the coordinates of `source` (geographic by
#' default): project the result yourself for plotting, e.g. with
#' [rproj_xy()] or your tool of choice.
#'
#' Points that cannot be projected (for example beyond the horizon of an
#' orthographic projection) do not error: segments touching them are left
#' un-densified, and refinement stops at unprojectable midpoints. Clipping
#' is out of scope here.
#'
#' @param x a two-column matrix of lon,lat (a path), a list of such
#'   matrices, or a segment mesh: a list with `vb` (4 x n vertex matrix,
#'   rows x, y, z, h) and `is` (2 x m segment index matrix, 1-based), as
#'   used by rgl's mesh3d
#' @param target projection for which to densify (proj string, WKT,
#'   authority code - anything PROJ accepts)
#' @param source coordinate system of the input, default 'OGC:CRS84'
#'   (longitude, latitude); authority codes are axis-normalized so
#'   'EPSG:4326' input is still given as lon,lat
#' @param tolerance maximum allowed deviation, in units of `target`
#'   (usually metres); if `NULL` (the default) it is derived from the
#'   projected extent of the input as `min(diff(range))/pixels`
#' @param pixels notional output width/height used to derive the default
#'   `tolerance`, i.e. "keep error under about one part in `pixels` of the
#'   plot"; ignored when `tolerance` is supplied
#' @param max_depth maximum bisection depth per input segment (each level
#'   at most doubles the vertex count of a segment)
#'
#' @return the same kind of object as `x`, with vertices added: a matrix
#'   for a path, a list of matrices for a list, a mesh with appended
#'   vertices and refined `is` for a mesh
#' @export
#'
#' @references <https://bost.ocks.org/mike/example/>
#' @examples
#' s <- segment(c(140, -164), c(-89, 80))
#' d <- densify(s, "+proj=laea +lon_0=147 +lat_0=-42")
#' nrow(d)
#' plot(rproj_xy(d, "+proj=laea +lon_0=147 +lat_0=-42"), type = "l", asp = 1)
densify <- function(x, target, source = "OGC:CRS84",
                    tolerance = NULL, pixels = 2048L, max_depth = 16L) {
  stopifnot(is.numeric(pixels), pixels >= 1, is.numeric(max_depth))
  max_depth <- as.integer(max_depth)

  if (is.matrix(x)) {
    tol <- tolerance %||% default_tolerance(x, target, source, pixels)
    out <- densify_path_cpp(x[, 1L, drop = TRUE], x[, 2L, drop = TRUE],
                            source, target, tol, max_depth)
    return(cbind(out$x, out$y))
  }

  if (is.list(x) && !is.null(x$vb) && !is.null(x$is)) {
    verts <- t(x$vb[1:2, , drop = FALSE])
    tol <- tolerance %||% default_tolerance(verts, target, source, pixels)
    out <- densify_mesh_cpp(verts[, 1L], verts[, 2L],
                            as.integer(x$is[1L, ]) - 1L,
                            as.integer(x$is[2L, ]) - 1L,
                            source, target, tol, max_depth)
    x$vb <- rbind(out$x, out$y, 0, 1)
    x$is <- rbind(out$s0 + 1L, out$s1 + 1L)
    return(x)
  }

  if (is.list(x) && all(vapply(x, is.matrix, logical(1L)))) {
    tol <- tolerance %||%
      default_tolerance(do.call(rbind, x), target, source, pixels)
    return(lapply(x, densify, target = target, source = source,
                  tolerance = tol, pixels = pixels, max_depth = max_depth))
  }

  stop("'x' must be a matrix, a list of matrices, or a segment mesh (vb/is)")
}

## default tolerance: one part in 'pixels' of the smaller side of the
## projected bounding box of the input vertices (the d3 screen-space idea)
default_tolerance <- function(verts, target, source, pixels) {
  xy <- rproj_xy(verts, target, source = source)
  ok <- is.finite(xy[, 1L]) & is.finite(xy[, 2L])
  if (!any(ok)) {
    stop("no input vertex is projectable in 'target'; supply 'tolerance' explicitly")
  }
  side <- min(diff(range(xy[ok, 1L])), diff(range(xy[ok, 2L])))
  if (!is.finite(side) || side <= 0) {
    stop("degenerate projected extent; supply 'tolerance' explicitly")
  }
  side / pixels
}

File: src/densify.cpp
#include <string>
#include <vector>

#include <cpp11.hpp>

#include <proj.h>

#include "bigcurve.h"

using namespace cpp11;

namespace {

// RAII PROJ adapter: one normalized PJ for the lifetime of a densify call.
class ProjTransform : public bigcurve::Transform {
public:
  ProjTransform(const std::string& from, const std::string& to) {
    PJ_CONTEXT* ctx = PJ_DEFAULT_CTX;
    PJ* raw = proj_create_crs_to_crs(ctx, from.c_str(), to.c_str(), NULL);
    if (raw == NULL) {
      stop("Error creating transform: %s",
           proj_context_errno_string(ctx, proj_context_errno(ctx)));
    }
    // always lon/lat in, easting/northing out, regardless of authority
    // axis order (so EPSG:4326 behaves like OGC:CRS84 here)
    PJ* norm = proj_normalize_for_visualization(ctx, raw);
    if (norm != NULL) {
      proj_destroy(raw);
      trans_ = norm;
    } else {
      trans_ = raw;
    }
  }

  ~ProjTransform() { proj_destroy(trans_); }

  bool fwd(double lon, double lat, double* x, double* y) override {
    PJ_COORD in;
    in.xyzt.x = lon;
    in.xyzt.y = lat;
    in.xyzt.z = 0.0;
    in.xyzt.t = HUGE_VAL;
    PJ_COORD out = proj_trans(trans_, PJ_FWD, in);
    *x = out.xy.x;
    *y = out.xy.y;
    return std::isfinite(*x) && std::isfinite(*y);
  }

private:
  PJ* trans_;
};

writable::doubles to_doubles(const std::vector<double>& v) {
  writable::doubles out((R_xlen_t)v.size());
  for (R_xlen_t i = 0; i < (R_xlen_t)v.size(); i++) out[i] = v[(size_t)i];
  return out;
}

writable::integers to_integers(const std::vector<int>& v) {
  writable::integers out((R_xlen_t)v.size());
  for (R_xlen_t i = 0; i < (R_xlen_t)v.size(); i++) out[i] = v[(size_t)i];
  return out;
}

}  // namespace

// Densify an ordered path of lon/lat vertices. Returns list(x, y) of the
// densified path in geographic coordinates, endpoints preserved.
[[cpp11::register]]
list densify_path_cpp(doubles lon, doubles lat,
                      std::string from, std::string to,
                      double tol, int max_depth) {
  ProjTransform t(from, to);
  bigcurve::Resampler rs(t, tol, max_depth);

  R_xlen_t n = lon.size();
  std::vector<bigcurve::Vertex> v((size_t)n);
  for (R_xlen_t i = 0; i < n; i++) {
    v[(size_t)i] = rs.project(lon[i], lat[i]);
  }

  std::vector<double> ox, oy;
  ox.reserve((size_t)n * 2);
  oy.reserve((size_t)n * 2);
  std::vector<bigcurve::Vertex> interior;

  for (R_xlen_t i = 0; i + 1 < n; i++) {
    ox.push_back(v[(size_t)i].lon);
    oy.push_back(v[(size_t)i].lat);
    interior.clear();
    rs.edge(v[(size_t)i], v[(size_t)i + 1], interior);
    for (size_t k = 0; k < interior.size(); k++) {
      ox.push_back(interior[k].lon);
      oy.push_back(interior[k].lat);
    }
  }
  if (n > 0) {
    ox.push_back(v[(size_t)(n - 1)].lon);
    oy.push_back(v[(size_t)(n - 1)].lat);
  }

  writable::list out({to_doubles(ox), to_doubles(oy)});
  out.names() = {"x", "y"};
  return out;
}

// Densify a segment mesh: vertices (x, y) and 0-based edge index pairs
// (s0, s1). Returns list(x, y, s0, s1, parent) with new vertices appended,
// each input edge replaced by its refined chain (0-based indices), and
// parent giving the 0-based input edge each output edge descends from --
// the hook for carrying per-segment attributes (feature, ring, arc)
// through refinement.
[[cpp11::register]]
list densify_mesh_cpp(doubles x, doubles y,
                      integers s0, integers s1,
                      std::string from, std::string to,
                      double tol, int max_depth) {
  ProjTransform t(from, to);
  bigcurve::Resampler rs(t, tol, max_depth);

  R_xlen_t nv = x.size();
  std::vector<bigcurve::Vertex> v((size_t)nv);
  for (R_xlen_t i = 0; i < nv; i++) {
    v[(size_t)i] = rs.project(x[i], y[i]);
  }

  std::vector<double> vx(x.begin(), x.end());
  std::vector<double> vy(y.begin(), y.end());
  std::vector<int> e0, e1, parent;
  e0.reserve((size_t)s0.size() * 2);
  e1.reserve((size_t)s0.size() * 2);
  parent.reserve((size_t)s0.size() * 2);
  std::vector<bigcurve::Vertex> interior;

  for (R_xlen_t k = 0; k < s0.size(); k++) {
    int a = s0[k];
    int b = s1[k];
    if (a < 0 || b < 0 || a >= (int)nv || b >= (int)nv) {
      stop("edge index out of range");
    }
    interior.clear();
    rs.edge(v[(size_t)a], v[(size_t)b], interior);
    int prev = a;
    for (size_t j = 0; j < interior.size(); j++) {
      vx.push_back(interior[j].lon);
      vy.push_back(interior[j].lat);
      int idx = (int)vx.size() - 1;
      e0.push_back(prev);
      e1.push_back(idx);
      parent.push_back((int)k);
      prev = idx;
    }
    e0.push_back(prev);
    e1.push_back(b);
    parent.push_back((int)k);
  }

  writable::list out({to_doubles(vx), to_doubles(vy),
                      to_integers(e0), to_integers(e1),
                      to_integers(parent)});
  out.names() = {"x", "y", "s0", "s1", "parent"};
  return out;
}
